### ISSUE TITLE
Exception is thrown on first error; doesn't collect list of validation errors

### ISSUE THREAD

**Issue by heddn** [drupalassociation/php-signify#4]
When looking at the following code, we throw an exception and bale immediately a problem occurs. It is more helpful if we could provide a list of issues and collect all the problems. Yes, the signature failed, but we shouldn't have to fix that before we see how many other issues exist.

```php
    public function verifyChecksumList($signed_checksum_list, $working_directory)
    {
        $checksum_list_raw = $this->verifyMessage($signed_checksum_list);
        return $this->verifyTrustedChecksumList($checksum_list_raw, $working_directory);
    }
    protected function verifyTrustedChecksumList($checksum_list_raw, $working_directory) {
        $checksum_list = $this->parseChecksumList($checksum_list_raw, true);
        $verified_count = 0;
        /**
         * @var VerifierFileChecksum $file_checksum
         */
        foreach ($checksum_list as $file_checksum)
        {
            $actual_hash = @hash_file(strtolower($file_checksum->algorithm), $working_directory . DIRECTORY_SEPARATOR . $file_checksum->filename);
            if ($actual_hash === false) {
                throw new VerifierException("File \"$file_checksum->filename\" in the checksum list could not be read.");
            }
            if (empty($actual_hash) || strlen($actual_hash) < 64) {
                throw new VerifierException("Failure computing hash for file \"$file_checksum->filename\" in the checksum list.");
            }
            if (strcmp($actual_hash, $file_checksum->hex_hash) !== 0)
            {
                throw new VerifierException("File \"$file_checksum->filename\" does not pass checksum verification.");
            }
            $verified_count++;
        }
        return $verified_count;
    }
```

**Comment by pwolanin** [comment]
That's a ton more code to write to collect the errors and less robust than halting on any error.

**Comment by heddn** [comment]
Yes it is. In automatic updates, I (we???)  want to show all the files that have failed hash validation. Not just the first one, then the 2nd one, etc. This is just a PoC of one way to collect the data.

**Comment by mbaynton** [comment]
@pwolanin @davidstrauss and I had several intentional discussions at MWDS around failing fast when it came to verifying keys and message signatures, but I think once those things are okay it's fine to collect all the incorrect hashes.

**Comment by davidstrauss** [comment]
I talked with @heddn on Drupal Slack, and he offered some excellent reasons why we should consider returning a report rather than throwing an exception on the first mismatch. I'll take a look at pull request #5 now.

**Comment by heddn** [comment]
After extensive conversation in slack, it sounds like the design of the project is to only validate downloaded files. Not to check already downloaded files. The two scenarios are subtly different.

In the first, we download then for security concerns make sure what we just downloaded is what we expect.
The second scenario takes a look at what is already on the filesystem and compares against a manifest file of known hashes. We want to validate the manifest hasn't changed. But running the rest of the checks from this library are not part of the design. That should be done independently.

**Comment by davidstrauss** [comment]
> After extensive conversation in slack, it sounds like the design of the project is to only validate downloaded files. Not to check already downloaded files. The two scenarios are subtly different.

We discussed these use cases a bit at MWDS, and our conclusion was similar to yours. That is, security validation is primarily for new data acquired over the network. This wasn't so much because we intended to target that use case, though. It came more from agreement that, if the anchor of trust is a public key on disk, then that has limited usefulness for validating the security of other files on disk (as an attacker could change both).

So, we ended up with the same view I think you have, which is to separate the concerns of data re-validation on disk versus validating untrusted data from the network. One thing we talked about was simply bundling the sha512sum file in tarballs and zip files.

> But running the rest of the checks from this library are not part of the design. That should be done independently.

Perhaps we could share some code and formats around validating the hash lists, though?
